### ISSUE TITLE
Multilinear operators

### ISSUE THREAD

**Issue by pmli** [pymor/pymor#304]
Multilinear operators (tensors) appear in quadratic systems. If a linear system is
```
x' = A x + B u,
y = C x,
```
a quadratic system would be
```
x' = A x + H^(1) (x kron x) + B u,
y = C x.
```
where `H^(1)` is the mode-1 matricization of a tensor `H` and `kron` is the Kronecker product. (`H` is the Hessian of `A x + H^(1) (x kron x)`, if symmetric, although it can be symmetrized in general.)

`H^(1)` is a matrix, but it's not a linear operator with respect to `x` (it's linear with respect to `x kron x`, but Kronecker product is expensive for high-dimensional vectors and is not available in pyMOR). So, it would make sense to add multilinear operators to pyMOR.

Interface class for multilinear operators would need to define mode-`mu` product methods. For example, `H^(1) (x kron x)` can be represented as `H` multiplied by `x` in mode-2 and mode-3. Quadratic system reduced by Petrov-Galerkin projection (assuming `W^T V = I`) is
```
x_r' = W^T A V x_r + W^T H^(1) (V kron V) (x_r kron x_r) + W^T B u,
y_r = C V x_r.
```
`W^T H^(1) (V kron V)` can be computed as `H` multiplied by `W^T` in mode-1, by `V^T` in mode-2, and by `V^T` in mode-3.

Multilinear operators could also be used to represent low-rank tensor formats, which is also in some sense model reduction.

@sdrave @ftalbrecht @renemilk I would like to hear your thoughts on adding multilinear operators to pyMOR.

**Comment by sdrave** [comment]
Generally speaking, we should definitely support multilinear operators in some form. Two possible options I see without having thought too much about it:

1. Add a new Interface class `MultilinearOperator`. `apply` would take a tuple of `VectorArrays` instead of a single `VectorArray`.
2. Multilinear operators are linear operators on the tensor product of their argument spaces. So the `source` of a multilinear operator could be appropriately chosen. (I would be fun to have a `TensorProductVectorArray`, but I am not sure how useful that would be.)

**Comment by pmli** [comment]
Option 1 seems better to me. Personally, I like to think about multilinear operators as multi-dimensional arrays. Also, we would not need to think about `apply_adjoint`, one `apply` method would be enough.

An implementation using NumPy should be straightforward. For a sparse representation, the situation is a bit difficult, since it seems that SciPy doesn't have sparse multi-dimensional arrays. Other options I found, besides doing it by hand, are [scikit-tensor](https://github.com/mnick/scikit-tensor/blob/master/sktensor/sptensor.py) (GPL 3 license) and [TensorFlow](https://www.tensorflow.org/versions/r0.11/api_docs/python/sparse_ops.html) (Apache 2 license). The former is "an extremely young project" (I'm citing the readme file), but it might be enough. The latter is big, maybe too big for our purposes.

For the Tensor Train (TT) format, there is [ttpy](https://github.com/oseledets/ttpy) (MIT license). Here, we would also need appropriate vector array class.

**Comment by sdrave** [comment]
I do not think that `apply_adjoint` is an argument since many methods of `OperatorInterface` are not required to be actually implemented (e.g. there is no `apply_adjoint` for nonlinear operators). Maybe it would help the discussion if we would try to list the main operations which need to be performed for multilinear operators (and for what).

Quadratic nonlinear operators appear, for instance, in the Navier-Stokes and Euler equations, and I assume that there are not many solvers around which would assemble the operator as a 3-tensor. Thus, it would be nice if we could define an interface that does not depend on any concrete matrix representation of the operator.

**Comment by pmli** [comment]
> Maybe it would help the discussion if we would try to list the main operations which need to be performed for multilinear operators (and for what).

1. For simulation, we would need H<sup>(1)</sup> (x &otimes; x). This could be done as `H.apply((None, x, x))` (multiplication in mode-2 and mode-3). The result should be a vector array (or easily transformed to a vector array). ODE solvers that use Hessians would be good to have, I guess.
2. When projecting, the "reduced H" is W<sup>T</sup> H<sup>(1)</sup> (V &otimes; V). This could be done as `H.apply((W, V, V))` and the result should be a reduced 3-tensor.
3. For TQB-IRKA \[1\] (Alg. 3.1): 
  1. the reduced H needs a coordinate transformation: R<sup>-1</sup> H<sup>(1)</sup> (R &otimes; R) (this might be better done without an extra interface method, e.g. by `scipy.linalg.solve(R, H.apply((None, R, R)).data)`)
  2. compute (with original H) `H.apply((None, V, V))` and `H.apply((W, V, None))` (a vector array should be the result)
4. For truncated Gramians (I can't add a reference, the work is not yet published), to solve corresponding Lyapunov equations, same operations as in 3.ii are necessary.

> Quadratic nonlinear operators appear, for instance, in the Navier-Stokes and Euler equations

The magical thing is that nonlinear systems consisting of elementary functions (polynomials, exponentials, trigonometric functions...) can be exactly transformed to a quadratic-bilinear system, simply by adding some more variables (the problem can be if too many variables have to be added) \[1\].

> it would be nice if we could define an interface that does not depend on any concrete matrix representation of the operator

I agree. For some particular PDEs, I guess it would make sense to represent the multilinear operator as a collection of linear operators (corresponding to different matricizations).

\[1\] Benner, Goyal, Gugercin: "H<sub>2</sub>-Quasi-Optimal MOR for QB Control Systems", [arXiv preprint](https://arxiv.org/abs/1610.03279)

**Comment by sdrave** [comment]
I see. In case `H` comes from a PDE solver, it should usually be possible to evaluate H.apply((None, V, W)) or `H.apply((U, V, W))`. However, `H.apply((V, W, None))` or `H.apply((V, None, W))` would require additional (non-standard) implementation work. 

How would `H` be typically be given for your applications? As the some matrization of the tensor I would suspect?

**Comment by pmli** [comment]
I don't have applications yet, but I guess `H` could be given elementwise or in a mode-1 matricization.
